File: php/src/Metadata/EntityMapJson.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Metadata;

use SimpleOrm\Json\CanonicalWriter;

/**
 * Exports an {@see EntityMap} as the conformance JSON defined in
 * spec/metadata-model.md, via {@see CanonicalWriter} (never `json_encode` —
 * CODING-STANDARD §8). The export is deliberately column-centric and
 * language-neutral: column names, neutral type tokens, and SQL-side parameter
 * names — never PHP property names, which legitimately differ per
 * implementation. Every port must produce byte-identical output for the same
 * entity; ported from the C# reference's `EntityMapJson`.
 */
final class EntityMapJson
{
    private function __construct()
    {
    }

    public static function export(EntityMap $map): string
    {
        $document = [
            'entity' => $map->entityName(),
            'source' => self::source($map),
            'key' => [
                'strategy' => $map->keyStrategy->value,
                'columns' => array_map(static fn (PropertyMap $p): string => $p->columnName, $map->keyProperties),
            ],
        ];

        if ($map->versionProperty !== null) {
            $document['version'] = $map->versionProperty->columnName;
        }

        $document['columns'] = self::columns($map);

        if ($map->indexes !== []) {
            $document['indexes'] = self::indexes($map);
        }

        if ($map->relationships !== []) {
            $document['relationships'] = self::relationships($map);
        }

        return CanonicalWriter::write($document);
    }

    /** @return array<string, mixed> */
    private static function source(EntityMap $map): array
    {
        $source = ['kind' => $map->kind->value];

        if ($map->kind === RelationKind::Statement) {
            $source['sql'] = self::normalizeSql((string) $map->definingSql);
            $source['parameters'] = self::parameters($map->statementParameters);

            return $source;
        }

        $source['name'] = $map->relationName;
        if ($map->schema !== null) {
            $source['schema'] = $map->schema;
        }

        if ($map->definingSql !== null) {
            $source['sql'] = self::normalizeSql($map->definingSql);
        }

        if ($map->kind === RelationKind::Procedure) {
            $source['parameters'] = self::parameters($map->statementParameters);
        }

        return $source;
    }

    /**
     * @param list<StatementParameter> $parameters
     * @return list<array{name: string, type: string}>
     */
    private static function parameters(array $parameters): array
    {
        return array_map(
            static fn (StatementParameter $p): array => ['name' => $p->name, 'type' => $p->type->value],
            $parameters,
        );
    }

    /** @return list<array<string, mixed>> */
    private static function columns(EntityMap $map): array
    {
        $columns = [];
        foreach ($map->properties as $property) {
            $entry = [
                'column' => $property->columnName,
                'type' => self::typeToken($property),
                'nullable' => $property->nullable,
            ];
            if ($property->key) {
                $entry['key'] = true;
            }

            if ($property->generated) {
                $entry['generated'] = true;
            }

            $columns[] = $entry;
        }

        return $columns;
    }

    /**
     * Neutral type tokens (spec/metadata-model.md). Unlike the C# reference,
     * which derives the token from the CLR type at export time, PHP's loaders
     * resolve `ColumnType` once at load time (CODING-STANDARD §10), so export
     * only reads it — except `Custom`, which carries the PHP type name instead
     * of the spec's `clr:<full name>` (CODING-STANDARD §10).
     */
    private static function typeToken(PropertyMap $property): string
    {
        if ($property->type !== ColumnType::Custom) {
            return $property->type->value;
        }

        return 'php:' . ($property->phpType ?? 'mixed');
    }

    /** @return list<array<string, mixed>> */
    private static function indexes(EntityMap $map): array
    {
        $indexes = [];
        foreach ($map->indexes as $index) {
            $entry = [
                'name' => $index->name,
                'columns' => array_map(
                    static fn (IndexColumn $c): array => ['column' => $c->columnName, 'direction' => $c->descending ? 'desc' : 'asc'],
                    $index->columns,
                ),
            ];
            if ($index->unique) {
                $entry['unique'] = true;
            }

            $indexes[] = $entry;
        }

        return $indexes;
    }

    /** @return list<array<string, mixed>> */
    private static function relationships(EntityMap $map): array
    {
        return array_map(static function (RelationshipMap $relationship) use ($map): array {
            /** @var array<string, mixed> */
            return match ($relationship->kind) {
                RelationshipKind::ManyToOne => [
                    'kind' => 'many_to_one',
                    'foreignKeyColumns' => array_map(
                        static function (string $n) use ($map): string {
                            $property = $map->property($n);
                            assert($property !== null, "many_to_one FK property '{$n}' must be mapped");

                            return $property->columnName;
                        },
                        $relationship->foreignKeyProperties,
                    ),
                    'references' => self::shortName($relationship->targetType),
                ],
                RelationshipKind::OneToOne, RelationshipKind::OneToMany => [
                    'kind' => $relationship->kind === RelationshipKind::OneToOne ? 'one_to_one' : 'one_to_many',
                    'references' => self::shortName($relationship->targetType),
                    'targetForeignKeyProperties' => array_map(self::pascalCase(...), $relationship->foreignKeyProperties),
                ],
                RelationshipKind::ManyToMany => [
                    'kind' => 'many_to_many',
                    'references' => self::shortName($relationship->targetType),
                    'through' => self::shortName((string) $relationship->linkType),
                    'linkForeignKeysToOwner' => array_map(self::pascalCase(...), $relationship->linkForeignKeysToOwner),
                    'linkForeignKeysToTarget' => array_map(self::pascalCase(...), $relationship->linkForeignKeysToTarget),
                ],
            };
        }, $map->relationships);
    }

    /**
     * PascalCases a PHP property name for the cross-language `targetForeignKeyProperties` /
     * `linkForeignKeys*` export fields (CODING-STANDARD §10: an interim rule — the spec
     * should switch these fields to columns, ADR-0026 — that keeps the export byte-identical
     * to the C# reference's own property names).
     */
    private static function pascalCase(string $propertyName): string
    {
        return ucfirst($propertyName);
    }

    /** Collapses whitespace so the exported SQL is layout-independent across implementations. */
    private static function normalizeSql(string $sql): string
    {
        $parts = preg_split('/\s+/u', trim($sql));

        return implode(' ', array_filter($parts !== false ? $parts : [], static fn (string $s): bool => $s !== ''));
    }

    private static function shortName(string $fqcn): string
    {
        $slash = strrpos($fqcn, '\\');

        return $slash === false ? $fqcn : substr($fqcn, $slash + 1);
    }
}

File: go/orm/internal/metadata/entity_map_json.go
package metadata

import (
	"strings"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
)

// Export renders m as the conformance JSON of spec/metadata-model.md — the
// exact port of EntityMapJson.cs, using core.CanonicalJSON so the byte output
// matches the C# reference's System.Text.Json encoding. Column-centric and
// language-neutral: column names, neutral type tokens, and SQL-side parameter
// names only — never Go field names, except in the interim
// targetForeignKeyProperties/linkForeignKeysTo* fields (CODING-STANDARD §10,
// ADR-0027), which carry core.ToPascalCase(fieldName) until the spec switches
// them to column names.
func Export(m *core.EntityMap) string {
	root := core.JSONObject{}
	root = root.Set("entity", m.EntityName())
	root = root.Set("source", buildSource(m))
	root = root.Set("key", buildKey(m))

	if m.VersionProperty != nil {
		root = root.Set("version", m.VersionProperty.ColumnName)
	}

	root = root.Set("columns", buildColumns(m))

	if len(m.Indexes) > 0 {
		root = root.Set("indexes", buildIndexes(m))
	}

	if len(m.Relationships) > 0 {
		root = root.Set("relationships", buildRelationships(m))
	}

	return core.CanonicalJSON(root)
}

func buildSource(m *core.EntityMap) core.JSONObject {
	source := core.JSONObject{}
	source = source.Set("kind", m.Kind.Token())

	if m.Kind == core.RelationStatement {
		source = source.Set("sql", normalizeSQL(m.DefiningSQL))
		source = source.Set("parameters", buildParameters(m.StatementParameters))
		return source
	}

	source = source.Set("name", m.RelationName)
	if m.Schema != "" {
		source = source.Set("schema", m.Schema)
	}
	if m.DefiningSQL != "" {
		source = source.Set("sql", normalizeSQL(m.DefiningSQL))
	}
	if m.Kind == core.RelationProcedure {
		source = source.Set("parameters", buildParameters(m.StatementParameters))
	}
	return source
}

func buildKey(m *core.EntityMap) core.JSONObject {
	key := core.JSONObject{}
	key = key.Set("strategy", m.KeyStrategy.Token())
	columns := make([]any, len(m.KeyProperties))
	for i, k := range m.KeyProperties {
		columns[i] = k.ColumnName
	}
	key = key.Set("columns", columns)
	return key
}

func buildParameters(params []core.StatementParameter) []any {
	result := make([]any, len(params))
	for i, p := range params {
		obj := core.JSONObject{}
		obj = obj.Set("name", p.Name)
		obj = obj.Set("type", core.TypeToken(p.Type, core.ColumnTypeOf(p.Type)))
		result[i] = obj
	}
	return result
}

func buildColumns(m *core.EntityMap) []any {
	result := make([]any, len(m.Properties))
	for i, p := range m.Properties {
		obj := core.JSONObject{}
		obj = obj.Set("column", p.ColumnName)
		obj = obj.Set("type", core.TypeToken(p.Type, p.ColumnType))
		obj = obj.Set("nullable", p.IsNullable)
		if p.IsKey {
			obj = obj.Set("key", true)
		}
		if p.IsGenerated {
			obj = obj.Set("generated", true)
		}
		result[i] = obj
	}
	return result
}

func buildIndexes(m *core.EntityMap) []any {
	result := make([]any, len(m.Indexes))
	for i, index := range m.Indexes {
		obj := core.JSONObject{}
		obj = obj.Set("name", index.Name)
		columns := make([]any, len(index.Columns))
		for j, c := range index.Columns {
			column := core.JSONObject{}
			column = column.Set("column", c.ColumnName)
			direction := "asc"
			if c.Descending {
				direction = "desc"
			}
			column = column.Set("direction", direction)
			columns[j] = column
		}
		obj = obj.Set("columns", columns)
		if index.Unique {
			obj = obj.Set("unique", true)
		}
		result[i] = obj
	}
	return result
}

func buildRelationships(m *core.EntityMap) []any {
	result := make([]any, len(m.Relationships))
	for i, r := range m.Relationships {
		obj := core.JSONObject{}
		switch r.Kind {
		case core.RelationshipManyToOne:
			obj = obj.Set("kind", "many_to_one")
			obj = obj.Set("foreignKeyColumns", foreignKeyColumns(m, r.ForeignKeyProperties))
			obj = obj.Set("references", r.TargetType.Name())

		case core.RelationshipOneToMany, core.RelationshipOneToOne:
			kindToken := "one_to_many"
			if r.Kind == core.RelationshipOneToOne {
				kindToken = "one_to_one"
			}
			obj = obj.Set("kind", kindToken)
			obj = obj.Set("references", r.TargetType.Name())
			obj = obj.Set("targetForeignKeyProperties", pascalCaseNames(r.ForeignKeyProperties))

		default: // core.RelationshipManyToMany
			obj = obj.Set("kind", "many_to_many")
			obj = obj.Set("references", r.TargetType.Name())
			obj = obj.Set("through", r.LinkType.Name())
			obj = obj.Set("linkForeignKeysToOwner", pascalCaseNames(r.LinkForeignKeysToOwner))
			obj = obj.Set("linkForeignKeysToTarget", pascalCaseNames(r.LinkForeignKeysToTarget))
		}
		result[i] = obj
	}
	return result
}

// foreignKeyColumns resolves many_to_one's FK property names (properties of
// this entity) to their column names — this side's own export, so column
// names are the language-neutral contract (spec/metadata-model.md).
func foreignKeyColumns(m *core.EntityMap, propertyNames []string) []any {
	result := make([]any, len(propertyNames))
	for i, name := range propertyNames {
		if p := m.Property(name); p != nil {
			result[i] = p.ColumnName
		}
	}
	return result
}

func pascalCaseNames(names []string) []any {
	result := make([]any, len(names))
	for i, name := range names {
		result[i] = core.ToPascalCase(name)
	}
	return result
}

// normalizeSQL collapses whitespace so exported SQL is layout-independent across implementations.
func normalizeSQL(sql string) string {
	return strings.Join(strings.Fields(sql), " ")
}
